File: frontend/src/api/workspace.js
import { requestJson } from "./client";

export async function getTree(path = "") {
  return requestJson(`/workspace/tree?path=${encodeURIComponent(path)}`);
}

export async function readFile(path) {
  return requestJson(`/workspace/file?path=${encodeURIComponent(path)}`);
}

export async function writeFile(path, content) {
  return requestJson("/workspace/file", {
    method: "PUT",
    body: JSON.stringify({ path, content }),
  });
}

export async function createFile(path, isDir = false) {
  return requestJson("/workspace/file", {
    method: "POST",
    body: JSON.stringify({ path, is_dir: isDir }),
  });
}

export async function deleteFile(path) {
  return requestJson(`/workspace/file?path=${encodeURIComponent(path)}`, {
    method: "DELETE",
  });
}

export async function renameFile(oldPath, newPath) {
  return requestJson("/workspace/rename", {
    method: "POST",
    body: JSON.stringify({ old_path: oldPath, new_path: newPath }),
  });
}

File: backend/app/api/routes/workspace.py
from __future__ import annotations

from fastapi import APIRouter, Depends, Query

from app.api.deps import get_workspace_service
from app.api.routes.auth import get_current_user
from app.schemas.workspace import (
    CreateFileRequest,
    FileContent,
    FileNode,
    RenameRequest,
    WriteFileRequest,
)
from app.services.workspace_service import WorkspaceService

router = APIRouter(prefix="/workspace", tags=["workspace"])


@router.get("/tree", response_model=list[FileNode])
async def get_tree(
    path: str = Query("", description="Relative path in workspace root"),
    current_user: dict = Depends(get_current_user),
    workspace_service: WorkspaceService = Depends(get_workspace_service),
) -> list[FileNode]:
    return workspace_service.get_tree(path)


@router.get("/file", response_model=FileContent)
async def read_file(
    path: str = Query(..., description="Relative file path"),
    current_user: dict = Depends(get_current_user),
    workspace_service: WorkspaceService = Depends(get_workspace_service),
) -> FileContent:
    return workspace_service.read_file(path)


@router.put("/file", response_model=FileContent)
async def write_file(
    req: WriteFileRequest,
    current_user: dict = Depends(get_current_user),
    workspace_service: WorkspaceService = Depends(get_workspace_service),
) -> FileContent:
    return workspace_service.write_file(req.path, req.content)


@router.post("/file", response_model=FileNode)
async def create_file(
    req: CreateFileRequest,
    current_user: dict = Depends(get_current_user),
    workspace_service: WorkspaceService = Depends(get_workspace_service),
) -> FileNode:
    return workspace_service.create_file_or_dir(req.path, req.is_dir)


@router.delete("/file")
async def delete_file(
    path: str = Query(..., description="Relative file or directory path"),
    current_user: dict = Depends(get_current_user),
    workspace_service: WorkspaceService = Depends(get_workspace_service),
) -> dict[str, bool]:
    success = workspace_service.delete_file_or_dir(path)
    return {"success": success}


@router.post("/rename", response_model=FileNode)
async def rename_file(
    req: RenameRequest,
    current_user: dict = Depends(get_current_user),
    workspace_service: WorkspaceService = Depends(get_workspace_service),
) -> FileNode:
    return workspace_service.rename_file_or_dir(req.old_path, req.new_path)
